codebase/src/compiler/checker-inference.md

```diff
@@ -0,0 +1,245 @@
+second line the same as if the author had written:
+
+``` ts
+setup<string>({ initial() { return "last" } })
+```
+
+Meaning that the compiler then checks that
+`{ initial() { return "last" } }` is assignable to
+`{ initial(): string }`.
+
+### Walkthrough
+
+Type parameter inference starts off in `inferTypeArguments`, where
+the first step in type parameter inference is to get the type of all
+the arguments to the function whose parameters are being inferred. In
+the above example, the checker says that the type of
+`{ initial() { return "last" } }` is `{ initial(): string }`. This
+type is called the **source** type, since it is the source of
+inferences. It's matched with the parameter type `{ initial(): T }`.
+This is the **target** type -- it contains type parameters which are
+the target of the process.
+
+Type parameter inference is a pairwise walk of the two types, looking
+for type parameters in the target, matching them to corresponding
+types in the source. The type is walked structurally sort of like a tree
+is elsewhere in the compiler.
+
+1. `inferTypes` gets called on each source/target pair with
+   argument=source/parameter=target. There's only one pair here:
+   `{ initial(): string }` and `{ initial(): T }`.
+2. Since both sides are object types, `inferFromProperties` looks
+   through each property of the target and looks for a match in the
+   source. In this case both have the property `initial`.
+3. `initial`'s type is a signature on both sides
+   (`() => T/() => string`), so inference goes to `inferFromSignature`, which
+   recursively infers from the return type.
+4. Now the source/target pair is `T/string`. Since the source is a
+   lone type parameters, we add `string` to the list of candidates for
+   `T`.
+
+Once all the parameters have had `inferTypes` called on them,
+`getInferredTypes` condenses each candidate array to a single type,
+via `getUnionType` in this case. `T`'s candidates array is `[string]`,
+so `getUnionType` immediately returns `string`.
+
+### Other considerations
+
+#### Method of Combining Candidate Arrays
+
+Only inference to return types, `keyof T` and mapped type constraints
+(which are usually `keyof` too) produce a union. All other locations
+call the custom code `getCommonSupertype`, which more or less does
+what it says. Note that object types are always unioned together
+first, regardless of inference position.
+
+#### Interference Between Contextual Typing and Type Parameter Inference
+
+Type parameter inference actually operates in two passes. The first
+pass skips arguments that have contextually typed expressions so that
+if good inferences are found from other arguments, contextual typing
+can provide types to parameters of function expressions, which in turn
+may produce better return types. Then the second pass proceeds with
+all arguments.
+
+#### Inference Priorities
+
+Different positions have different inference priorities; when the type
```

Comment:
Maybe worth mentioning that priorities are fully heuristic in nature - there's no reason for a priority to exist other than it being useful at filtering possible inferences away in practice. A full unifying solver would try all the possible inferences and pick the one that works - instead, we essentially sort them all by "likelihood to work based on location found", and pick the top one, no questions asked. Honestly, it might not be complete, but it is fast.